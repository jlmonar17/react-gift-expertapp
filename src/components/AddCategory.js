import { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
    const [inputValue, setInputValue] = useState("");

    const handleChange = (e) => {
        setInputValue(e.target.value);
    };

    const handleSubmit = (e) => {
        e.preventDefault();

        if (inputValue.length > 2) {
            setCategories((category) => [inputValue, ...category]);
            setInputValue("");
        }
    };

    return (
        <form onSubmit={handleSubmit}>
            <input type="text" onChange={handleChange} value={inputValue} />
        </form>
    );
};

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
};
