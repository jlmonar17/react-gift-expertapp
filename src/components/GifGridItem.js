export const GifGridItem = ({ id, title, url }) => {
    return (
        <div className="card">
            <img src={url} alt={title} key={id} />
            <p>{title}</p>
        </div>
    );
};
