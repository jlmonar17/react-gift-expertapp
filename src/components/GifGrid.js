import { GifGridItem } from "./GifGridItem";
import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from "prop-types";

export const GifGrid = ({ category }) => {
    const { data: images, loading } = useFetchGifs(category);

    return (
        <>
            <h3 className="animate__animated animate__pulse">{category}</h3>

            {loading && (
                <p className="animate__animated animate__flash">Cargando...</p>
            )}

            <div className="card-grid">
                {images.map((image) => {
                    return <GifGridItem key={image.id} {...image} />;
                })}
            </div>
        </>
    );
};

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
};
