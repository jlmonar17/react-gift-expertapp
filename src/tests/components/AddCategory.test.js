import { shallow } from "enzyme";
import { AddCategory } from "../../components/AddCategory";

describe("Pruebas para <AddCategory />", () => {
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories} />);

    beforeEach(() => {
        wrapper = shallow(<AddCategory setCategories={setCategories} />);
        jest.clearAllMocks();
    });

    test("Debe mostrarse de manera correcta", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("Probamos que la caja de texto cambie correctamente su value al escribir texto", () => {
        const input = wrapper.find("input");
        const value = "Hola Mundo";

        input.simulate("change", {
            target: { value },
        });

        const inputAfter = wrapper.find("input");

        expect(inputAfter.prop("value")).toBe(value);
    });

    test("No se debe hacer submit del formulario cuando el input está vacío", () => {
        const form = wrapper.find("form");

        form.simulate("submit", { preventDefault: () => {} });

        // Se espera que la función setCategories no sea llamada cuando se le pasa un input vacio en el submit.
        expect(setCategories).not.toHaveBeenCalled();
    });

    test("Debe de llamarse el setCategories y dejar el input vacío", () => {
        const value = "Naruto";

        wrapper.find("input").simulate("change", { target: { value } });
        wrapper.find("form").simulate("submit", { preventDefault: () => {} });

        // set categories debe haberse llamado ya que se tiene un value
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1);
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

        // Input debe quedar vacío luego de llamarse el setCategories
        const inputAfter = wrapper.find("input");
        expect(inputAfter.prop("value")).toBe("");
    });
});
