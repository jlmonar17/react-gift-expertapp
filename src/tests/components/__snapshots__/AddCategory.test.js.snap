// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas para <AddCategory /> Debe mostrarse de manera correcta 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
